ergo-lib/src/wallet/multi_sig.rs

```diff
@@ -0,0 +1,844 @@
+            .collect();
+        let secret_bag = HintsBag::empty();
+        let public_bag = HintsBag::empty();
+        let old_secret: &Vec<Hint> = &self.secret_hints.get(&index).unwrap_or(&secret_bag).hints;
+        for hint in old_secret {
+            secret.push(hint.clone());
+        }
+
+        let old_public: &Vec<Hint> = &self.public_hints.get(&index).unwrap_or(&public_bag).hints;
+        for hint in old_public {
+            public.push(hint.clone());
+        }
+        self.secret_hints.insert(index, HintsBag { hints: secret });
+        self.public_hints.insert(index, HintsBag { hints: public });
+    }
+
+    /// Outputting HintsBag corresponding for an index
+    pub fn all_hints_for_input(&self, index: usize) -> HintsBag {
+        let mut hints: Vec<Hint> = Vec::new();
+        let secret_bag = HintsBag::empty();
+        let public_bag = HintsBag::empty();
+        let secrets: &Vec<Hint> = &self.secret_hints.get(&index).unwrap_or(&secret_bag).hints;
+        for hint in secrets {
+            hints.push(hint.clone());
+        }
+        let public: &Vec<Hint> = &self.public_hints.get(&index).unwrap_or(&public_bag).hints;
+        for hint in public {
+            hints.push(hint.clone());
+        }
+        let hints_bag: HintsBag = HintsBag { hints };
+        hints_bag
+    }
+}
+
+/// Compute commitments for UncheckedLeaf
+pub fn compute_commitments(leaf: UncheckedLeaf) -> Option<FirstDlogProverMessage> {
+    let mut ret: Option<FirstDlogProverMessage> = None;
+
+    match leaf {
+        UncheckedLeaf::UncheckedSchnorr(pdlog) => {
+            let challenge = pdlog.challenge;
+            let proposition = pdlog.proposition;
+            let second_message = pdlog.second_message;
+            let comm = compute_commitment(&proposition, &challenge, &second_message);
+            ret = Some(FirstDlogProverMessage::from(comm));
+        }
+        UncheckedLeaf::UncheckedDhTuple(_) => {}
+    }
+    ret
+}
+
+#[allow(clippy::unwrap_used)]
+/// Outputs Hints bag for a signed(invalid) transaction
+pub fn bag_for_multi_sig(
+    sigma_tree: SigmaBoolean,
+    real_propositions: &[SigmaBoolean],
+    simulated_propositions: &[SigmaBoolean],
+    proof: &[u8],
+) -> HintsBag {
+    let ut: UncheckedTree = parse_sig_compute_challenges(&sigma_tree, proof.to_owned()).unwrap();
+    let mut bag: HintsBag = HintsBag::empty();
+    traverse_node(
+        ut,
+        real_propositions,
+        simulated_propositions,
+        NodePosition::crypto_tree_prefix(),
+        &mut bag,
+    );
+    bag
+}
+
+#[allow(clippy::unwrap_used)]
+/// Generate commitments for transaction
+pub fn generate_commitments(
+    tx_context: TransactionContext,
+    state_context: &ErgoStateContext,
+    public_keys: &[SigmaBoolean],
+) -> Result<TransactionHintsBag, TxSigningError> {
+    let mut hints_bag = TransactionHintsBag::empty();
+    for (i, input) in tx_context.get_boxes_to_spend().enumerate() {
```

Comment:
`tx_context.get_boxes_to_spent()` are not the same as tx inputs. Although there must be a box in `boxes_to_spend` for every tx inputs the boxes might be in a different order.

Reply:
Yes, I didn't know that was fixed. i handled it in the new commit.